Workaround in onnx to get transposes into init_nets

zdevito opened the pull request:
This adds a pass to ONNX so that it can speculate Transpose
operators so that ONNX's split pass can put them into an init_net

zdevito committed:
Workaround in onnx to get transposes into init_nets

This adds a pass to ONNX so that it can speculate Transpose
operators so that ONNX's split pass can put them into an init_net

Also fixes a potential bug in onnx peephole where an optimization
across blocks might move a Value and violate scoping.

jamesr66a commented on torch/csrc/jit/passes/onnx/peephole.cpp:
```diff
@@ -373,6 +377,36 @@ void fixDefaultLstmCellState(Block *b) {
+      continue;
+    // find the control flow node in the same block as node_input that contains
+    // Node n
+    auto flow_node = n->owningBlock()->owningNode();
```
can you name this `control_flow_node`? :p

jamesr66a commented on test/test_jit.py:
```diff
@@ -2833,6 +2833,22 @@ def foo(x, y):
+            Foo(),
+            (torch.ones(2, 1, dtype=torch.float), ),
+            None, verbose=False)
+        self.assertExpected(onnx_ish)
```
Where's the expect file?

apaszke commented on torch/csrc/jit/passes/onnx/peephole.cpp:
```diff
@@ -373,6 +377,36 @@ void fixDefaultLstmCellState(Block *b) {
+    // put the node right before this flow node
+    n->moveBefore(flow_node);
+  }
+}
```
What's the purpose of this? Are we adding it because it helps with optimization? Maybe we should do the same for regular JIT codepaths?

zdevito replied:
We need to move transposes of weights into the ONNX `init_net` for perf reasons. But things inside control flow will not move to the init_net just the top-level nodes. ONNX Transpose is safe to speculate (it just reverses dimensions). Torch tranpose is not (it checks dim==2, i think). Moving it outside of an if statement can cause errors in a previously correct program. Almost _every_ operator in the graph is unsafe to speculate in this way, so this pass is very specific.

apaszke replied:
(Just a general comment, not to this PR) I'm not sure why the transpose in torch would be unsafe, since you know that it's input won't change because of the blocks that you take it out of.

apaszke reviewed (commented):
LGTM, but needs to have the submodule updates removed.

jamesr66a commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -359,6 +359,7 @@ void PropagateShapeOnBlock(Block * block) {
       setDynamicType(node);
     } catch(std::exception & e) {
       if(auto sl = node->getSourceLocation()) {
+        std::cout << *block->owningGraph() << "\n";
```
You probably want to delete this

jamesr66a commented on test/test_jit.py:
```diff
@@ -2833,6 +2833,34 @@ def foo(x, y):
+            @torch.jit.script_method
+            def forward(self, x):
+                x += x
+                if True:
```
How about make one of these levels of control a loop?

zdevito replied:
y has to be assigned along all paths, so putting it into a loop would require putting `y = self.m(x)` before the loop but that would make it difficult to observe that the transpose call gets lifted.

zdevito committed:
Perform shape propagation when embedding a program into a trace.

This ensures the trace still has type information specific to that trace, which will help onnx export succeed in more cases.

ezyang reviewed (approved):
Speculation algorithm looks good.

ezyang commented:
Missing a Windows guard

```
20:03:35 ======================================================================
20:03:35 ERROR: test_trace_of_script (test_jit.TestScript)
20:03:35 ----------------------------------------------------------------------
20:03:35 Traceback (most recent call last):
20:03:35   File "C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test\test\test_jit.py", line 2633, in test_trace_of_script
20:03:35     self.assertEqual(3, use(torch.ones(1, dtype=torch.float)))
20:03:35 RuntimeError: NYI: fuser is not supported on Windows.
20:03:35 
20:03:35 ----------------------------------------------------------
```

zdevito commented on torch/csrc/jit/export.cpp:
```diff
@@ -261,7 +261,7 @@ void encodeBlock(onnx::GraphProto * p_g, Block *b,
       body->set_type(onnx::aGRAPH);
       auto g = body->mutable_g();
-      encodeBlock(g, node->blocks()[0], initializers, ctx, raw_data_export_map);
+      encodeBlock(g, node->blocks()[0], {}, ctx, raw_data_export_map);
```
that was a bug - those initializers were only for the top level block and they should not be propagated to subblocks
